### ISSUE TITLE
Implement `pre-commit-hooks` in rust

### ISSUE THREAD

**Issue by j178** [j178/prek#17]

**Comment by wshanks** [comment]
It would be nice to document how the builtin hooks work. All I could find documented was that some hooks have builtin Rust implementations. From looking at the code, I believe what happens is that you still specify the https://github.com/pre-commit/pre-commit-hooks/ url, but for any hook ids with builtin versions the builtin version gets used instead? Additionally, perhaps if only builtin ids are specified the pre-commit-hooks repo is not even cloned?

**Comment by j178** [comment]
@wshanks Here you go https://prek.j178.dev/builtin/

**Comment by wshanks** [comment]
Thanks! I didn't see an answer to the part about whether `prek install` downloads the repo if you only use builtin hooks (haven't had a chance to test) so I am guessing it does. It might be a possible future extension to allow setting the url to "builtin" or something like that. I was interested in using `prek` in an environment without access to GitHub (I have other workaround options but was wondering if I could get by without them).

**Comment by j178** [comment]
> whether prek install downloads the repo if you only use builtin hooks

Yeah, for now, prek only swaps the implementation to Rust, but we still pull hook info like name and types from the remote repo. So, we do need to clone it first.

> allow setting the url to "builtin" or something like that

That's a good idea, I'll think about it.

**Comment by mlisovyi** [comment]
I would be also interested in setting the URL (e.g. via and env variable) would be indeed helpful in an environment without github access.

One could use a fork of that repo in a private on-prem gitlab instance (for example) or elsewhere, but that leads to hooks not being matched because of the URL and default python hooks being used instead of the built-ins (one sees it in the TRACE logging)

@wshanks what is the workaround that you mention?

**Comment by leiserfg** [comment]
@j178  I just crated a PR to use `prek` as a replacement of `pre-commit` in https://github.com/cachix/git-hooks.nix. But there, `pre-commit-hook` is not fetched from github, we use the one already available in nixpkgs. Is there any way (or could it be added) to hint prek so it uses the builtin regardless?

**Comment by j178** [comment]
@leiserfg Hi, the builtin check logic is here: https://github.com/j178/prek/blob/master/src/builtin/mod.rs#L16-L30. I'm open to adding support for other pre-commit hook sources, but I don't really know much about nix - this would probably need help from the community.

**Comment by leiserfg** [comment]
For a bit more of clarity, [git-hooks.nix](https://github.com/cachix/git-hooks.nix) generates a  `.pre-commit-config.yaml` in the repo whith this tructure:

```yaml
{
  "default_stages": [
    "pre-commit"
  ],
  "repos": [
    {
      "hooks": [
...   # all the hooks in use
],
   "repo": "local"
    }
  ]
}
```

And there, every hook is defined with an "entry" that points to the absolute path of a binary in nixpkgs, so `pre-commit` and `prek` don't do any fetching, they just executed the hooks. I would like to mark the hooks that currently use the `pre-commit-hook` executable, in a way that when using `prek` it would not call the `pre-commit-hook` binary, but the builtin implementation instead.


Maybe using the fast-path also for `local` could do it, but that might cause problems if someone is using a local hook with the same id as the ones in pre-commit-hook repo, so maybe we better get read of it 🤔. Another option could be to use url = `local`,    and id = `builtin-original-hook-id` (like    `builtin-end-of-file-fixer`) but not sure if that's too hacky.

**Comment by wshanks** [comment]
> @wshanks what is the workaround that you mention?

Sorry, @mlisovyi, the workarounds that I was thinking of would use the Python versions of the hooks and not the builtin ones. The two options I was thinking of were either forking the pre-commit-hooks repo to a local copy (so the url would not match and the builtin hooks would not be used) or using local/script hook with the pre-commit-hooks PyPI package as a dependency (requires a little more work to re-define the hooks but avoids using `git` entirely).

**Comment by leiserfg** [comment]
@j178  what do you think of adding a new optional key to the hooks, (false by default) that tells if the hook should be replaced by a builtin one with the same id? That would work with our use case. My point is not to replace the current heuristic, just to also match those with the new key in true.

**Comment by j178** [comment]
@leiserfg I'm hesitant to add a new configuration key for this. I'm working on adding a special repo key called "builtin" that would include all the currently implemented builtin hooks. For example:

```yaml
repos:
  - repo: builtin
    hooks:
     - id: trailing-whitespace 
```

By adding this special "builtin" repo, we wouldn't need to clone anything or create hook environments like we do with `pre-commit/pre-commit-hooks` - it'd work completely offline without any network connectivity.

**Comment by leiserfg** [comment]
@j178 can you consider using the already existing `local` one? That way it will actually work for us without any  change and will be backward compatible with pre-commit.

**Comment by j178** [comment]
@leiserfg In the original pre-commit framework design, `local` isn’t that different from remote hooks - it just skips the repo cloning but still has to set up the hook environment (creating virtualenvs, installing dependencies, etc.). Right now, there’s no way for `local` to indicate that it needs a special built-in implementation.

> it will actually work for us without any change

What does your configuration look like when you use “repo: local” to get pre-commit-hooks running?

**Comment by leiserfg** [comment]
Within https://github.com/cachix/git-hooks.nix we don't use the external repos, we put all in `local` because we take care of fetching the binaries from nixpkgs ourselves, see one example of it here:

[kfz70fqbyw7jbn1dlg6fnfp39hdlfwk5-pre-commit-config.json](https://github.com/user-attachments/files/22763485/kfz70fqbyw7jbn1dlg6fnfp39hdlfwk5-pre-commit-config.json)

**Comment by j178** [comment]
@wshanks @mlisovyi Just a heads‑up: `repo: builtin` was added in #1118 and will be included in the next release.

**Comment by j178** [comment]
> Within https://github.com/cachix/git-hooks.nix we don't use the external repos, we put all in `local` because we take care of fetching the binaries from nixpkgs ourselves, see one example of it here:

@leiserfg I think I’m starting to get how git‑hooks.nix works now, but I still don’t have a clear idea of how to make it play nicely with prek's built‑in hooks when using `repo: local`. We can talk it through in a new issue if that’s necessary.
